Separate the task runner and the command line parser

w33ble opened the pull request:
Blocked by #24
This PR starts at 62c0e60

This PR decouples the command parser from the task runner, and exposes the task runner when it's `require`d in node applications.

Doing so allows consumers to execute tasks without calling out the command line or going through a command line parser.

w33ble committed:
remove conditionl require in runner

add tasks module to load all tasks, throw if task does not exist

w33ble committed:
change task signatures and action handlers

- tasks get plugin, run, and an options object
- action handler controls what goes into each task
- taskRunner wrapper moves the command object to the first argument

w33ble committed:
change test command signature

allow files to be passed in, and pass all options to test:server and test:browser

w33ble committed:
simplify the task runner

w33ble committed:
fix typo in unknownOptions

w33ble committed:
expose the task runner as the module's main

this way tasks can be run programatically without going through a cli parser

spalger commented on bin/plugin-helpers.js:
```diff
@@ -39,11 +39,13 @@ program
 
 program
-  .command('test')
+  .command('test [files...]')
```
This doesn't feel right. It's always going to run the browser tests and then filter the server tests... lets just make this a shortcut for the defaults unless there is a good reason otherwise.

spalger commented on bin/plugin-helpers.js:
```diff
@@ -15,33 +23,53 @@ enableCollectingUnknownOptions(
 
 program
-  .command('test')
+  .command('test [files...]')
```
This doesn't feel right. Lets just keep `test` a shortcut for "all test files, browser and server"

w33ble committed:
add tests for task runner

w33ble committed:
remove file passing for testAll

w33ble committed:
add serverTestPaths to the plugin config

useful for overriding the value via a config file

w33ble commented:
Removed the options passing for the testAll task. Moved the location of the server tests files into the plugin config, so it can be overridden with the config files.

spalger committed:
[config] plugin.serverTestPaths -> plugin.serverTestPatterns

spalger commented:
![image](https://cloud.githubusercontent.com/assets/1329312/21255866/c7104c32-c32b-11e6-86ca-ed939c4f674c.png)


Changed the new config value to `plugin.serverTestPatterns` to reflect their globyness
